src/source_map.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use crate::module_specifier::ModuleSpecifier;
+
+use deno_ast::ParsedSource;
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde_json::from_slice;
+use serde_json::Value;
+use sourcemap::decode_data_url;
+use sourcemap::DecodedMap;
+
+lazy_static! {
+  static ref SOURCE_MAP_URL_RE: Regex =
+    Regex::new(r#"(?i)^# sourceMappingURL=([^\s'"]+)"#).unwrap();
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum ParsedSourceMap {
+  Value(Value),
+  Url(ModuleSpecifier),
+}
+
+fn find_source_map_url(parsed_source: &ParsedSource) -> Option<String> {
+  for comment in parsed_source.comments().trailing_map().values() {
+    for item in comment.iter().rev() {
+      if let Some(caps) = SOURCE_MAP_URL_RE.captures(&item.text) {
+        if let Some(m) = caps.get(1) {
+          return Some(m.as_str().to_string());
+        }
+      }
+    }
+  }
+  None
+}
+
+/// Attempt to identify and parse a source map URL from a source file, returning
+/// the parsed source map (if inlined) or the specifier to the source map.
+pub(crate) fn parse_sourcemap(
+  specifier: &ModuleSpecifier,
+  parsed_source: &ParsedSource,
+) -> Option<ParsedSourceMap> {
+  let source_map_url = find_source_map_url(parsed_source)?;
+  let source_map_specifier =
+    if let Ok(url) = ModuleSpecifier::parse(&source_map_url) {
+      Some(url)
+    } else if let Ok(url) = specifier.join(&source_map_url) {
+      Some(url)
+    } else {
+      None
+    }?;
```

Comment:
I think this could maybe use `or_else` here:

```rs
let source_map_specifier = ModuleSpecifier::parse(&source_map_url)
  .or_else(|| specifier.join(&source_map_url))
  .ok()?;
```